Create sideset extrusion generator

RocksSavage opened the pull request:
## Reason
<!--Why do you need this feature or what is the enhancement?-->
Adds a SidesetExtruderGenerator, to extrude meshes out of their sidesets. 

## Design
<!--A concise description (design) of the enhancement.-->
This is a wrapper for 4 other generators, applied in succession. This is a shortcut generator, if you will.


## Impact
<!--Will the enhancement change existing APIs or add something new?-->
The users will be happy :)
Easier interface for users for sideset extrusion

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

RocksSavage committed:
Add a SideSetExtruderGenerator and related documentation to the supporting sub-generators.

ref #20880

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1108882/) on a93c841 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21573/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21573/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d655d3a3e0548088bc3138295662d810c5f0702a`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit e1211aaf2dec766e9a8e2da1631ee2f36ec7c5de
+Subproject commit 42c944f1e3d97126bc519733f0f9cda93fa7042b
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1108951/) on 0b09708 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21573/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1109977/) on 95507b6 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d655d3</th>
      <th colspan=4>#21573 95507b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.54%</td>
      <td>83.55%</td>
      <td>+0.01%</td>
      <td>96.10%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>77882</td>
      <td>77955</td>
      <td>+73</td>
      <td>74</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15340</td>
      <td>15344</td>
      <td>+4</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21573/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21573/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21573/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/doc/content/source/meshgenerators/SideSetExtruderGenerator.md:
```diff
@@ -0,0 +1,65 @@
+
+The SideSetExtruder Generator is for "pulling" mesh in a direction. Given a sideset and a direction(vector), it adds to the mesh. SideSetExtruder differs from [MeshExtruderGenerator](MeshExtruderGenerator.md) in that the extrusion stays in the same dimension as the sideset it is pulling on (MeshExtruderGenerator is for pulling 2 dimensional shapes into 3 dimensions). 
+
+SideSetExtruder will not throw any errors if you extrude a mesh through (collide with) another mesh. It will throw an error if the extrusion vector would create a mesh with a negative determinant (or rather, [MeshExtruderGenerator](MeshExtruderGenerator.md) throws the error). The extrusion vector is applied relative to the sideset given, not from the origin point of the mesh. The extrusion vector ...
```
```suggestion
SideSetExtruder will not throw any errors if you extrude a mesh through (collide with) another mesh. It will throw an error if the extrusion vector would create a mesh with a negative determinant (or rather, the underlying subgenerator [MeshExtruderGenerator](MeshExtruderGenerator.md) throws the error). The extrusion vector is applied relative to the sideset given, not from the origin point of the mesh. The extrusion vector must be a 3D vector even if you are only extruding a 2D mesh; in such a case, let the z component be 0 (e.g., `extrusion_vector = '1 0.5 0'`)
```

GiudGiud commented on framework/doc/content/source/meshgenerators/StitchedMeshGenerator.md:
```diff
@@ -15,6 +15,8 @@ Consider the following three meshes.
 Using the `StitchedMeshGenerator` object from within the [Mesh](/Mesh/index.md) block of the input file, as shown in the input
 file snippet below, these three square meshes are joined into a single mesh as shown in Figure 4.
 
+Note that the way that the meshes are merged gives precedence to the left-most mesh listed in terms of sidesets: the sidesets of the second, third, etc meshes will be subsumed into the sidesets of the first mesh. The names of the sidesets in the first mesh are what the names that will remain in the outputted mesh. 
```
```suggestion
Note that the way that the meshes are merged gives precedence to the left-most mesh listed in terms of sidesets: the sidesets of the second, third, etc meshes will be absorbed into the sidesets of the first mesh. The names of the sidesets in the first mesh are what the names will remain in the outputted mesh. 
```

RocksSavage replied:
Neither of us did that last sentence justice. It should be "The names of the sidesets in the first mesh are what names will remain in the outputted mesh. "

Thanks for the other suggestion. Will implement.

GiudGiud commented on framework/include/meshgenerators/BlockToMeshConverterGenerator.h:
```diff
@@ -0,0 +1,31 @@
+  /// Mesh that comes from another generator
+  std::unique_ptr<MeshBase> & _input;
+
+  const std::vector<SubdomainName> _target_blocks;
```
add docstring

GiudGiud commented on framework/include/meshgenerators/SideSetExtruderGenerator.h:
```diff
@@ -0,0 +1,36 @@
+  /// Mesh that comes from another generator
+  std::unique_ptr<MeshBase> * _build_mesh;
+
+  const MeshGeneratorName _original_input;
+  const RealVectorValue _extrusion_vector;
+  const unsigned int _num_layers;
+  const BoundaryName _sideset_name;
```
add docstrings

GiudGiud commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+  InputParameters params = MeshGenerator::validParams();
+
+  params.addClassDescription(
+      "Converts a block (subdomain) from a mesh into a stand-alone mesh with one block in it.");
```
```suggestion
      "Converts a block (subdomain) from a mesh into a stand-alone mesh.");
```

GiudGiud commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+  params.addRequiredParam<MeshGeneratorName>("input", "The mesh we want to modify");
+  params.addRequiredParam<std::vector<SubdomainName>>(
+      "target_blocks",
+      "The (list of) blocks (or 'subdomains') we wish to have moved to a new mesh (by name, not "
+      "ID)");
```
```suggestion
      "The (list of) blocks (or 'subdomains') we wish to have moved to a new mesh");
```

passing ids should work fine here actually

GiudGiud commented on framework/src/meshgenerators/SideSetExtruderGenerator.C:
```diff
@@ -0,0 +1,100 @@
+    _sideset_name(getParam<BoundaryName>("sideset"))
+{
+  // constants needed only temporarily
+  const BoundaryName _EXTRUDED_BLOCK_NAME = "extruded_block";
+  const BoundaryName _SIDESET_TO_BE_STITCHED = "to_be_stitched";
```
use regular moose naming convention:
extruded_block_name 
No capitals, no leading underscore since not a class member

GiudGiud reviewed (dismissed):
love the extensive testing and documentation. 

@roystgnr could you please review this?
I'm thinking with lowerDBlockGenerator + the BlockToMeshGenerator in this PR we will be able to make a 'WithinSidesetTriangulator'

roystgnr commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  if (!mesh->is_replicated())
+    mooseError("BlockToMeshConverterGenerator is not implemented for distributed meshes.");
```
Would it be sufficient to just throw on a `MeshSerializer` here?  "Doesn't work efficiently" is still a step up from "doesn't work at all".

roystgnr replied:
Well, there's support for distributed meshes, and then there's "support" for distributed meshes.  Temporarily serializing a mesh is an awful excuse for support ... but it's better than nothing, especially since proper distributed mesh support can be a *lot* of work.  In the context of a mesh generator temporary serialization often makes sense, when users have a workflow where they construct a small (and hence cheap to serialize) mesh at one stage, and only refine or extrude or stitch or whatever to turn it into a large (and hence necessary to distribute) mesh at a later stage.

GiudGiud replied:
This is a good idea.

We likely should remove all these blanket mooseError statements and add serializers in all the MGs instead as a first step towards having fully distributed generation.

roystgnr commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+
+          // Add to mesh. This method call will automatically re-assign the id; so must call this
+          // first
+          new_mesh->add_node(node_copy.get());
```
This is kind of dangerous, leaving two `unique_ptr` pointing to the same target and then manually releasing the latter.  It's also kind of weird, manually creating the Node yourself without setting an id.  Why not just `Node * node_copy = new_mesh->add_point(elem->point(i));`?

RocksSavage replied:
>This is kind of dangerous, leaving two unique_ptr pointing to the same target and then manually releasing the latter. 

The `.get()` method returns, not another unique_ptr -- but a raw pointer of what the unique_ptr is owning. 

> It's also kind of weird, manually creating the Node yourself without setting an id. 

I thought that the `add_node` method would automatically assign the id, correctly. When I tried to assign it myself it didn't work for me. From what I gathered, the id gets used as the index of the node in the list kept internally by the mesh. Therefore the id must be unique, and ordered correctly. The `i` variable loops through the list of nodes in each element, even while the elements were being looped over. There was no way for me to choose what order the elements were looped over, to ensure that the point I was `add_point`-ing had an id that was in ascending order in the new mesh. 

Hope that made some kind of sense. There could be something I'm not understanding here. Is it important to keep the node id's consistent from one mesh to the next?

> Why not just `Node * node_copy = new_mesh->add_point(elem->point(i));`?

That is an elegant alternative. I'm way pre-maturely optimizing here, but I speculated that just re-assigning from the ptr would be more efficient, since it wouldn't require creating a new variable, which would take up another register in the CPU. (probably silly, but that was my rationale)

roystgnr replied:
> The .get() method returns, not another unique_ptr -- but a raw pointer of what the unique_ptr is owning.

Sure.  And when you pass that raw pointer to `add_node`, you're promising us that you allocated it (true!) and that it's our responsibility to delete it (false! total lie! you've still got another deleter assigned to it!).  Our fault, for not documenting that well enough, but it's still a code smell.  You manually release the redundant deleter afterward, but the whole point of smart pointers is to make it so those dangerous situations provably never happen.

> I thought that the add_node method would automatically assign the id, correctly.

"Correctly" is a matter of context, but in the context here it'll do fine, giving you an id one past the maximum existing id.  My point is that that's *also* what `add_point` would do, with less work in your code.  You just get the `Node *` returned by `add_point` and check that to see what `id()` it got assigned.  You've got 4 lines of code here that could be replaced by 3; that's not a problem because that 4th line is going to take up too much hard drive space or anything, but just because it's misleading, it suggests that something special is being done here.

... likewise for `copy_n_index`, actually, now that I look at that.  Isn't that always going to be the same as `i`?

> I'm way pre-maturely optimizing here

Yeah.  `add_node` and `add_point` are both virtual function calls (they're inserting into different data structures depending on whether your mesh is `ReplicatedMesh` or `DistributedMesh`); even if your code did less work that would be dwarfed by the virtual thunk and stack manipulation.

roystgnr commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+
+      // it is ok to release the copy element into the mesh because derived meshes class
+      // (ReplicatedMesh, DistributedMesh) manage their own elements, will delete them
+      new_mesh->add_elem(copy.release());
```
Just `add_elem(copy)`; libMesh will take a `unique_ptr` here.

In general, using `unique_ptr::release` is a code smell; it discards the safety guarantees you would otherwise get by using smart pointers as intended.

RocksSavage replied:
I checked out a copy of the branch and made the change to see if it would work

```

In file included from /Users/savatc/projects/moose_v2/framework/build/unity_src/meshgenerators_Unity.C:5:
/Users/savatc/projects/moose_v2/framework/src/meshgenerators/BlockToMeshConverterGenerator.C:119:26: error: call to implicitly-deleted copy constructor of 'std::unique_ptr<Elem>'
      new_mesh->add_elem(copy);
                         ^~~~

```

It appears that the copy constructor of the unique_ptr class has been purposefully deleted, to prevent people from making a copy of a unique_ptr. This is to reinforce the notion that there should only be one unique_ptr pointing at an object, I reckon. 

The reason I `release()`-ed the unique_ptr here was because the unique would pass out of scope on the next line (120), and the the destructor of the unique_ptr would be called, which would delete the `Elem` object I had just added to the mesh. Then the mesh would have a bunch of null pointers to `Elem`s that no longer existed. 

The `release` method returns, not another unique_ptr --  but a raw pointer of what the unique_ptr is owning.

roystgnr replied:
Sorry, I meant to say `add_elem(std::move(copy))`.  You're right, you do have to explicitly tell the compiler that you're doing a move rather than a copy of the parameter.

roystgnr commented on framework/src/meshgenerators/SideSetExtruderGenerator.C:
```diff
@@ -0,0 +1,100 @@
+std::unique_ptr<MeshBase>
+SideSetExtruderGenerator::generate()
+{
+  return dynamic_pointer_cast<MeshBase>(*_build_mesh);
```
Why is the cast necessary here?  Isn't `*_build_mesh` already a `unique_ptr<MeshBase>`?

RocksSavage replied:
I'll fix that, thanks!

Accidentally implemented _build_mesh as a raw pointer pointing at a unique_ptr instead of reference of one

GiudGiud committed:
Misc fixups:
- remove all caps variables
- avoid overlapping names for sidesets and blocks
- clearer docs
- more docstrings
refs #20880

GiudGiud commented:
Should be ready for another round or merge @roystgnr

roystgnr reviewed (dismissed):
Couldn't find any non-trivial issues remaining.

GiudGiud commented on test/tests/meshgenerators/block_to_mesh_converter_generator/tests:
```diff
@@ -12,7 +12,8 @@
       recover = false
       cli_args = '--mesh-only'
-      mesh_mode = 'replicated'
+      mesh_mode = 'REPLICATED'      
```
no it's not. Trent must have seen upper case somewhere else

GiudGiud replied:
cleaned up

GiudGiud commented on framework/src/meshgenerators/SideSetExtruderGenerator.C:
```diff
@@ -0,0 +1,97 @@
+    params.set<std::vector<SubdomainName>>("target_blocks") = {extruded_block_name};
+
+    // convert lower dimensional block to a separate mesh
+    _build_mesh =
```
well it's more copy-pastable that way

GiudGiud replied:
like if someone steals our sub-generators it will be viable in all positions the way it is here

GiudGiud commented:
@roystgnr should be good for re-approval please

GiudGiud commented:
Thanks for the prompt reviews!
